appscale/tools/node_layout.py

```diff
@@ -367,202 +317,259 @@ def is_valid_simple_format(self):
+      if role == 'master' and master_node is not None:
+        self.invalid("Only one master is allowed")
 
-        if role == 'database':
-          # The first database node is the master
-          if index == 0:
-            is_master = True
-          else:
-            is_master = False
-          node.add_db_role(is_master)
-        elif role == 'db_master':
-          node.add_role('zookeeper')
-          node.add_role('role')
-        elif role == 'taskqueue':
-          # Like the database, the first taskqueue node is the master
-          if index == 0:
-            is_master = True
-          else:
-            is_master = False
-          node.add_role('taskqueue')
-          node.add_taskqueue_role(is_master)
-        else:
-          node.add_role(role)
-        
-        node_hash[ip_addr] = node
+      # Create or retrieve the nodes from the node_hash.
+      nodes = [node_hash[ip] if ip in node_hash else AdvancedNode(ip, cloud)
+               for ip in ips_list]
 
-    # Dont need the hash any more, make a nodes list
-    nodes = node_hash.values()
+      # Validate volume usage, there should be an equal number of volumes to
+      # number of nodes.
+      if self.disks:
+        self.validate_disks(nodes)
 
-    for node in nodes:
-      if not node.is_valid():
-        return self.invalid(",".join(node.errors()))
+        for node in nodes:
+          node.disk = self.disks.get(node.public_ip)
 
-      if self.infrastructure in InfrastructureAgentFactory.VALID_AGENTS:
-        if not self.NODE_ID_REGEX.match(node.public_ip):
-          return self.invalid("{0} is not a valid node ID (must be node-int)".
-            format(node.public_ip))
-      else:
-        # Virtualized cluster deployments use IP addresses as node IDs
-        if not self.IP_REGEX.match(node.public_ip):
-          return self.invalid("{0} must be an IP address".format(
-            node.public_ip))
+      # Add the defined roles to the nodes.
+      for node in nodes:
+        node.add_role(role)
+
+      # Check cases where a master is needed.
+      if role == 'master':
+        master_node = nodes[0]
+      if role == 'database' and not db_master_created:
+        nodes[0].add_db_role(True)
+        db_master_created = True
+      if role == 'taskqueue' and not tq_master_created:
+        # Check if we have more than one node to choose from and the first node
+        # is already the database master.
+        if 'db_master' in nodes[0].roles and len(nodes) > 1:
+          nodes[1].add_taskqueue_role(True)
+        else:
+          nodes[0].add_taskqueue_role(True)
+      if role == 'login':
+        login_found = True
+      elif role == 'login' and login_found:
+        self.invalid("Only one login is allowed")
 
-    if self.add_to_existing:
-      self.nodes = nodes
-      return self.valid()
+      # Update dictionary containing role counts.
+      role_count.update({(role, role_count.get(role, 0) + len(nodes))})
```

Comment:
agh, yes it should, I got my parantheses mixed up. the reason I was doing update was so that I didn't have to initialize the dictionary with roles besides the necessary ones.